net: lib: coap: Introduce net mgmt events for CoAP

pdgendt opened the pull request:
Allow users to register net mgmt events callbacks for CoAP events.
Currently CoAP service started/stopped and observer added/removed events will be notified.

pdgendt committed:
net: lib: coap: Introduce net mgmt events for CoAP

Allow users to register net mgmt events callbacks for CoAP events.

Signed-off-by: Pieter De Gendt <pieter.degendt@basalte.be>

pdgendt commented on subsys/net/lib/coap/coap_server.c:
```diff
@@ -433,6 +434,9 @@ int coap_service_start(const struct coap_service *service)
 
 	coap_server_update_services();
 
+	net_mgmt_event_notify_with_info(NET_EVENT_COAP_SERVICE_STARTED, NULL, service,
+					sizeof(struct coap_service));
```
Is it possible to know if the `sock_fd` is bound to a specific interface to be passed here?

pdgendt commented on include/zephyr/net/coap_mgmt.h:
```diff
@@ -0,0 +1,65 @@
+/* CoAP events */
+#define _NET_COAP_LAYER		NET_MGMT_LAYER_L4
+#define _NET_COAP_CODE		0x1c0
+#define _NET_COAP_IF_BASE	(NET_MGMT_EVENT_BIT |			\
+				 NET_MGMT_LAYER(_NET_COAP_LAYER) |	\
+				 NET_MGMT_LAYER_CODE(_NET_COAP_CODE))
```
What does the `NET_MGMT_IFACE_BIT` do, and should it be added here?

pdgendt commented on include/zephyr/net/coap_mgmt.h:
```diff
@@ -0,0 +1,65 @@
+
+/** @endcond */
+
+enum net_event_coap_cmd {
+	NET_EVENT_COAP_CMD_SERVICE_STARTED = 1,
+	NET_EVENT_COAP_CMD_SERVICE_STOPPED,
+};
```
I recently added `CONFIG_COAP_OBSERVER_EVENTS` for added/removed observers, would it be a better fit to have these events added here?
Or keep them both?

rlubos replied:
IMO it'd make sense to keep all of the CoAP-level events together. If we start to use net_mgmt framework for such events, IMO we should stick to it instead of having custom handler. So

pdgendt commented on include/zephyr/net/coap_mgmt.h:
```diff
@@ -0,0 +1,65 @@
+/**
+ * @brief coap_mgmt event raised when a service has started
+ */
+#define NET_EVENT_COAP_SERVICE_STARTED			\
+	(_NET_COAP_IF_BASE | NET_EVENT_COAP_CMD_SERVICE_STARTED)
```
Is this only for events, or should I also add start and stop commands to be used with `CONFIG_NET_MGMT`?

rlubos replied:
`NET_EVENT_*_CMD_*` could be also used to regsister `NET_REQUEST_*` and the request handler, I'm not sure if it's mandatory though. IMO in this case it's an unecessary overhead, the apps are better to use CoAP APIs directly and just stick to the events layer. Unless there's some usecase for `net_mgmt()` I don't see here.

pdgendt reviewed (commented):
I'd like some feedback on this PR to make sure I get the `NET_MGMT` part correct.
CC @rlubos @jukkar

pdgendt committed:
Revert "net: lib: coap: Add support for observer event callbacks"

This reverts commit 5227f248153f67e05b0ede9f859deb6f63ed4456.

The coap observer events will be replaced with net_mgmt events.

Signed-off-by: Pieter De Gendt <pieter.degendt@basalte.be>

pdgendt committed:
samples: net: sockets: coap_server: Remove observer events

In preparation of migrating to net_mgmt events, we need to remove
the observer event callback handler.

Signed-off-by: Pieter De Gendt <pieter.degendt@basalte.be>

pdgendt committed:
samples: net: sockets: coap_server: Add CoAP events example

Add an example handler for CoAP events triggered by net_mgmt.

Signed-off-by: Pieter De Gendt <pieter.degendt@basalte.be>

pdgendt committed:
doc: connectivity: networking: Update CoAP event handling example

Change the CoAP documentation to demonstrate CoAP events using
net_mgmt.

Signed-off-by: Pieter De Gendt <pieter.degendt@basalte.be>

pdgendt committed:
doc: release: 3.6: Add CoAP events documentation

Update the release notes and migration guide for CoAP events using
the Network Events subsystem.

Signed-off-by: Pieter De Gendt <pieter.degendt@basalte.be>

jukkar reviewed (commented):
Looks good. Could you add decoding of the events to net-shell events command in `subsys/net/lib/shell/events.c`, there are already some L4 events there so these can be added too.

pdgendt committed:
net: lib: shell: Add CoAP descriptions to event monitor

Add a description for each CoAP event on L4.

Signed-off-by: Pieter De Gendt <pieter.degendt@basalte.be>

rlubos reviewed (approved):
LGTM
